### PR TITLE
test(model): support fp32 with flash_attn

### PR DESCRIPTION
**Pull request by yingtongxiong**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

support fp32 with flash-attention

## Modification

InternLM/internlm/model/multi_head_attention.py
InternLM/internlm/initialize/launch.py

## Use cases (Optional)
企业微信截图_16927947592133
set the model config: dtype could be "torch.tf32" or "torch.float32"; use_flash_attn is True.

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.

### CODE REVIEW

File: internlm/model/multi_head_attention.py [hunk 1]
```diff
@@ -128,7 +128,13 @@ def _forward(self, x, seqlen=None, inference_params=None, **kwargs):
             qkv = self.rotary_emb(qkv, **kwargs)
 
         if inference_params is None:
-            context = self.inner_attn(qkv)
+            if gpc.config.model.dtype is torch.float32 and gpc.config.model.use_flash_attn:
+                with torch.cuda.amp.autocast(dtype=torch.float16):
+                    if qkv not in [torch.float16, torch.bfloat16]:
```
Comment (sunpengsdu) [inline internlm/model/multi_head_attention.py:133]: qkv .dtype ?

Comment (yingtongxiong) [reply internlm/model/multi_head_attention.py:133]: 哦哦，看错了。

Comment (yingtongxiong) [reply internlm/model/multi_head_attention.py:133]: 这里是用来判断，如果我model的dtype是float32且开了use_flash_attn，就对attention进行局部的autocast。qkv.dtype在133行判断了。 这样的话，如果model的dtype是float32，也能去走非flash_attn那条路。

Comment (sunpengsdu) [reply internlm/model/multi_head_attention.py:133]: 这里应该是qkv.dtype not in [torch.float16, torch.bfloat16] 吧？

Comment (yingtongxiong) [reply internlm/model/multi_head_attention.py:133]: 哦哦哦哦 好的

Comment (yingtongxiong) [reply internlm/model/multi_head_attention.py]: 已更新

**Pull request merged**